desc/objectives/_stability.py

```diff
@@ -367,3 +369,246 @@ def compute(self, params, constants=None):
+
+        eq = self.things[0]
+
+        if constants is None:
+            constants = self.constants
+        # we first compute iota on a uniform grid to get correct averaging etc.
+        iota_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._iota_keys,
+            params=params,
+            transforms=constants["iota_transforms"],
+            profiles=constants["iota_profiles"],
+        )
+
+        len_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._len_keys,
+            params=params,
+            transforms=constants["len_transforms"],
+            profiles=constants["len_profiles"],
+        )
+
+        # Now we compute theta_DESC for given theta_PEST
+        iota = iota_data["iota"][0]
+        rho, alpha, zeta = constants["fieldline_nodes"].T
+        theta_PEST = alpha + iota * zeta
+        nodes = jnp.array([rho, theta_PEST, zeta]).T
+
+        N_alpha = constants["N_alpha"]
+        N_zeta = constants["N_zeta"]
+
+        # Rootfinding theta for a given theta_PEST
+        desc_coords = map_coordinates(eq, nodes, inbasis=("rho", "theta_PEST", "zeta"))
+
+        import pdb
+
+        pdb.set_trace()
+
+        sfl_grid_list = [
+            Grid(
+                desc_coords[i * N_zeta : (i + 1) * N_zeta, :], sort=False, jitable=True
+            )
+            for i in range(N_alpha)
+        ]
+        transforms_list = [
+            get_transforms(self._data_keys, obj=eq, grid=sfl_grid, jitable=True)
+            for sfl_grid in sfl_grid_list
+        ]
+        profiles_list = [
+            get_profiles(self._data_keys, obj=eq, grid=sfl_grid)
+            for sfl_grid in sfl_grid_list
+        ]
+
+        # --no-verify sfl_grid = Grid(desc_coords[:N], sort=False, jitable=True)
+        # --no-verify transforms = get_transforms(
+        # --no-verify     self._data_keys, obj=eq, grid=sfl_grid, jitable=True
+        # --no-verify )
+        # --no-verify profiles = get_profiles(self._data_keys, obj=eq, grid=sfl_grid)
+
+        # we prime the data dict with the correct iota values so we don't recompute them
+        # using the wrong grid
+        data = {
+            "iota": iota_data["iota"][0] * jnp.ones_like(zeta),
+            "iota_r": iota_data["iota_r"][0] * jnp.ones_like(zeta),
+            "shear": iota_data["shear"][0] * jnp.ones_like(zeta),
+            "a": len_data["a"],
+        }
+
+        ## now compute ballooning stuff
+        # --no-verify data = compute_fun(
+        # --no-verify     "desc.equilibrium.equilibrium.Equilibrium",
+        # --no-verify     self._data_keys,
+        # --no-verify     params=params,
+        # --no-verify     transforms=transforms,
+        # --no-verify     profiles=profiles,
+        # --no-verify     data=data,
+        # --no-verify )
+
+        # now compute ballooning stuff
+        vectorized_compute = jnp.vectorize(
```

Comment:
I think you actually want to vectorize *inside* the compute function. Something like

```python
@register_compute_fun
...
def _ideal_bal_gamma(params, transforms, profiles, data, **kwargs):
    Nalpha = kwargs['Nalpha']
    Nzeta = kwargs['Nzeta']
    # get all the data you need and reshape it 
    newdata = {}
    newdata['B'] = data['B'].reshape((-1, Nalpha, Nzeta))
    newdata['g_rr'] = data['g_rr'].reshape((-1, Nalpha, Nzeta))
    # vectorize call over fieldlines, zeta is trailing 
    # dimension that doesn't get vectorized
    gamma = jnp.vectorize(_ideal_bal_gamma_onefieldline, signature="(n,)->()")(newdata)
    data['gamma'] = gamma
    return data

def _ideal_bal_gamma_onefieldline(data):
# this just computes gamma for a single field line, 
# all data is assumed to be 1d arrays along zeta, 
# basically all the stuff currently in ideal_bal_gamma2 
# but with the extra dimensions removed.

Reply:
actually looks like `jnp.vectorize` doesn't play nice with dicts, but you can use a double vmap:
```python
gamma = vmap(vmap(_ideal_bal_gamma_onefieldline))(newdata)

Reply:
Have you tested it yourself or is it a suggestion?

Reply:
suggestion, but if i understand your code correctly it should work. I can try implementing it for real. Is there a recommended test I can run to make sure I don't break things?

Reply:
Refactored the whole code.